app/models/concerns/two_factor_authentication_methods.rb

```diff
@@ -0,0 +1,27 @@
+module TwoFactorAuthenticationMethods
+  extend ActiveSupport::Concern
+
+  def piv_cac_configuration
+    PivCacConfiguration.new(self)
+  end
+
+  def auth_app_configuration
+    AuthAppConfiguration.new(self)
+  end
+
+  def personal_key_configuration
+    PersonalKeyConfiguration.new(self)
+  end
+
+  def two_factor_enabled?
+    two_factor_configurations.any?(&:mfa_enabled?)
+  end
+
+  def two_factor_configurations
+    phone_configurations + webauthn_configurations + [piv_cac_configuration, auth_app_configuration]
+  end
+
+  def total_mfa_options_enabled
+    two_factor_configurations.count(&:mfa_enabled?)
+  end
+end
```

Comment:
We generally shy away from model concerns, except when we have to, like in the Devise-related ones, or except in some cases where they're used in more than one model, like `NonNullUuid`. 

Here, this concern is only used by the User model, and all it does is move methods from the User model into the concern, which is like sweeping the dust under the rug :)

I would recommend extracting this logic into a dedicated class, such as a Policy Object.

Here is a great blog post on this subject:
https://blog.codeship.com/when-to-be-concerned-about-concerns/

> I am saying if you find yourself reaching to do a module-method extraction, stop. Consider what road you’re starting down and these lessons I’ve shared. What would a class-based extraction look like instead? If you’re not sure, there’s a wealth of object-oriented materials for purchase or via conference talks for free on YouTube. Use your best judgment and have fun.

And another classic article:
https://codeclimate.com/blog/7-ways-to-decompose-fat-activerecord-models/

> Don’t Extract Mixins from Fat Models
Let’s get this out of the way. I discourage pulling sets of methods out of a large ActiveRecord class into “concerns”, or modules that are then mixed in to only one model. I once heard someone say:
“Any application with an app/concerns directory is concerning.”
And I agree. Prefer composition to inheritance. Using mixins like this is akin to “cleaning” a messy room by dumping the clutter into six separate junk drawers and slamming them shut. Sure, it looks cleaner at the surface, but the junk drawers actually make it harder to identify and implement the decompositions and extractions necessary to clarify the domain model.

Reply:
Well, we can either have it in a module or in the class. I'm not going to write a policy object for something that's not a policy. I'm going to stick with the module for now rather than flap back and forth.

Reply:
When we say "policy", we refer to a business rule. In this case, the business rule that we are querying is whether or not the user is MFA-enabled. Please read these articles for more explanations and examples of Policy Objects:

https://medium.com/@steverob/policy-object-d158495265fa
https://pdabrowski.com/blog/ruby-on-rails/refactoring/policy-object-pattern

Having these methods in a module or in the User class is recognized as a bad practice.

Reply:
To clarify the main issue here, based on my experience, the books and blog posts I've read, the courses I've taken, and the conference talks I've watched, keeping business logic in an ActiveRecord model is considered a bad practice because it will hinder the maintainability of the project in the long run, and it invites more and more logic until the model becomes unwieldy. 

This is in fact exactly what happened 3 days ago when PR #2501 pushed the User model over the 100-line limit that we have set for ourselves. We use Rubocop to warn us when a class becomes too big, because that is a sign it is doing too much and that we should extract some parts into dedicated classes. Yet, instead of taking the Rubocop warning as an opportunity to reflect on the size of the User class and how we might make it leaner, we told Rubocop to ignore it, which goes against one of our [guidelines](https://github.com/18F/identity-idp/blob/master/CONTRIBUTING.md#style-readability-and-oo). And because the class size is now ignored by Rubocop, any additional methods we add to make it even bigger won't get automatically flagged.

Simply moving some methods out of the model and into a module that then gets included in the model doesn't solve the problem. It's still considered part of the ActiveRecord model, which is what we want to avoid. **Given that extracting the logic into a dedicated class is a straightforward process, and is known to make the code more maintainable, why would we not do it?**

The name of the new class is not the important issue here, it's that the User model has gotten too big and we need to extract some of it into a separate class. Since 2FA is a core concept in the app, that seems like a great fit for a new class, and because the main idea in this model concern is the method `two_factor_enabled?`, a Policy Object seems like a good fit. Perhaps something like this:

```ruby
class MfaEnabledUserPolicy
  def initialize(user)
    @user = user
  end
  
  def mfa_enabled?
    two_factor_configurations.any?(&:mfa_enabled?)
  end

  def total_mfa_options_enabled
    two_factor_configurations.count(&:mfa_enabled?)
  end

  private
  
  attr_reader :user
   
  def two_factor_configurations
    user.phone_configurations + user.webauthn_configurations + [piv_cac_configuration, auth_app_configuration]
  end

  def piv_cac_configuration
    PivCacConfiguration.new(user)
  end

  def auth_app_configuration
    AuthAppConfiguration.new(user)
  end
end
```
Again, the name of the class is not that important. If a Policy Object isn't our cup of tea, we could drop the `Policy` from the name, and put it in `app/models/mfa_enabled_user.rb` for example.

Reply:
We have to find a balance between code sprawl and readability. I don't want to have some mfa configuration access methods in one class (and have to explicitly instantiate that class to get them) and others on the user class. I want them all together so they're all accessible from the user model. OR all require another class wrap the user model. But since some configurations are active record relationships, that isn't possible.

I'm trying to hide some of the dirty details of how mfa configurations are managed so as we move away from the user model, we can get some consistency across the mfa methods.

Rigorous and blind following of dogma does not lead to more maintainable code. It has to be weighed against expressiveness and how we are trying to model the problem we're solving. Experienced developers can come in and follow the logic. They can see the pattern and know where to expect to find related but different cases.